docs: Extracting winners from a closed auction guide (#577)
paritytech/substrate-api-sidecar: AUCTIONS_WINNERS.md

<h1 style="text-align: center">WinnersData</h1>

## Summary

> This is a guide to leverage substrate-api-sidecar's endpoints in order to extract the necessary data to know the winner's of an auction based on its `auctionIndex`.

Reviewer: ```suggestion
In this guide we will learn how to track auction winners using [substrate-api-sidecar](https://github.com/paritytech/substrate-api-sidecar).
```

## Indexing

> One important note to finding the winner of each auction is making sure to index the block at which an auction ends. Since Sidecar is a stateless API, and due to the storage architecture for auctions in Substrate, you must index and keep track of certain fields during an ongoing auction so that the data collected can be used to find out which paraId's have won certain lease periods. (This will be explained further in the walkthrough).

Reviewer: ```suggestion
To find the winner of a completed auction we will need to know the block number the auction ended at. Since Sidecar is a stateless API and auction info is wiped once it is over we need the block number to make historic queries to a state when the relevant data was stored (keep reading for details).
```
Auctions are in polkadot, not substrate :)

Author: Ahhh yes, good catch! Also, I really Love the wording on this one.

> This can be done by leveraging the `/experimental/paras/auctions/current` endpoint. 

Reviewer: Can we clarify what "this" is in the above sentence

When there is an ongoing auction the return object will look like following below:

```
{
  "at": {
    "hash": "string",
    "height": "string"
  },
  "beginEnd": "string",
  "finishEnd": "string",
  "phase": "opening",
  "auctionIndex": "string",
  "leasePeriods": [
    "string"
  ],
  "winning": [
    {
      "bid": {
        "accountId": "string",
        "paraId": "string",
        "amount": "string"
      },
      "leaseSet": [
        "string"
      ]
    }
  ]
}
```

> Important keys here are `finishEnd` which is the last block at which the auction will take place, `auctionIndex` which will act as your point of reference for that specific auction, and `leasePeriods` which are the available lease period indexes that may be bid on for the specific `auctionIndex`. 

Reviewer: ```suggestion
We will index `finishEnd` (which is the last block at which the auction will take place), `auctionIndex` (which will act as a unique auction identifier), and `leasePeriods` (which are the available lease period indexes that may be bid on for the specific `auctionIndex`). 
```

Reviewer: Given the below we can probably just remove the parts in the `(..)`.

** Details of each key **

> `finishEnd`: This is the last block at which the auction will take place. Indexing it will allow you to query the block at which the auction ended. From that block you can extract the lease winning related events. (To query the block: GET `/blocks/{finishEnd}`.)

Reviewer: I am probably misusing the word - I just mean store in database

> `finishEnd`: This is the last block at which the auction will take place, and also the most important piece of data needed. This will allow you to query the block at which the auction has ended and will give you all the events that have taken place during that block. 

Reviewer: ```suggestion
`finishEnd`: This is the last block at which the auction will take place. This will allow you to query the block at which the auction has ended and will give you all the events that have taken place during that block. (To query the block: GET `/blocks/{finishEnd}`.)
```

> `auctionIndex`: Also known as the `AuctionCounter`, the index will be a pointer or ID to a specific auction. 

Reviewer: ```suggestion
`auctionIndex`: The unique identifier for the auction. 
```

> `leasePeriods`: These are the available lease period indexes that may be bid on for the specific `auctionIndex`. You can use this to create two seperate relationships explained below. 

Reviewer: ```suggestion
`leasePeriods`: The available lease period indexes that may be bid on for the specific `auctionIndex`. You can use this to create two separate relationships, as explained below. 
```
Wondering if there is a more descriptive way we can describe these "relationships"

Author: Yea I agree. I defaulted to "relationships" since it's shared language when referring to DB's but that being said that could also cause confusion. 

Spit balling some words => "association", "link", "connection".

Author: Agreed, just removed it.

## Relationships

With the information you are now tracking there are two main relationships to note. 

1. `auctionIndex` => `leasePeriods` (Lease Periods that are actively being auctioned off)

> 2. `paraId` => `leasePeriods` (Lease Periods that have been won by `paraId`).

Author: lower case is the way here you are correct, if we were mentioning its type though `LeasePeriod` then it would be capitalized. Probably where my confusion was.

> 1. `auctionIndex` => `leasePeriods` (Lease Periods that are actively being auctioned off)

Reviewer: ```suggestion
1. `auctionIndex` => `leasePeriods` (Lease Periods that are actively being auctioned off)
```

> 2. `paraId` => `leasePeriods` (Lease Periods that have been won).

Reviewer: ```suggestion
2. `paraId` => `leasePeriods` (Lease Periods that have been won by `paraId`).
```

## Walkthrough

### Relay-chain Dev Setup for auctions (SKIP if you already have a development setup)!
-------------

The begining of this guide will start by briefly introducing setting up a simple parachain enviornment if you dont already know how or dont have one set for local developement. 

> Start by cloning [polkadot](https://github.com/paritytech/polkadot) and checking out the `rococo-v1` branch. NOTE: Before compiling make sure to adjust the [EndingPeriod](https://github.com/paritytech/polkadot/blob/rococo-v1/runtime/rococo/src/lib.rs#L745) to `100`, and [LeasePeriod](https://github.com/paritytech/polkadot/blob/rococo-v1/runtime/rococo/src/lib.rs#L761) to `100` so the auction time is fit for local development. You can then follow this [gist](https://gist.github.com/emostov/a58f887fce6af8a9b4aa2421114836c5) to get your alice and bob dev validator nodes up. Then you can call those extrinsic calls with the polkadot-js UI. 

Reviewer: do we want rococo-v1 and not master? At a certain point I think it would make the most sense to use the polkadot release that the first set of real auctions will run on.

Author: Yea ill change it to polkadot master. When the release with the runtime that will contain the auctions comes out we should definitely update this to reflect that release like you mentioned.

### Using Sidecar to find the auction winners

`/experimental/paras/auctions/current` ENDPOINT

> An auction will either be in two phases `starting` or `ending`. During this period when querying the endpoint you will receive a `finishEnd` which will denote the last block where the `AuctionClosed` event will take place as well as the `Leased` event. These events will be under the `on_initialize` key. 

Reviewer: ```suggestion
An ongoing auction will be in one of two phases: `starting` or `ending`. During an ongoing auction the endpoint you will return a `finishEnd` which will denote the block where the `AuctionClosed` event and `Leased` events will be deposited. The aforementioned events will be deposited under the `on_initialize` key. 
```

Reviewer: Agreed, emitted does make sense more sense of the user perspective

Author: Awesome fixed this one up, and changed it to use an active voice.

> It is important to index this block for the current `auctionIndex` because this will be your source of truth of where the winners of the auction are stored. 

Reviewer: ```suggestion
By indexing the `finishEnd` block and looking at the `Leased` events within it, we can see whom the auction winners are and what lease periods they where rewarded.
```

lets either link to docs for leased event?

> The next important key is the `leasePeriods` that corresponds to the current `auctionIndex`. We will use these available `leasePeriods` to compare the winning results and see which `paraId`'s took which slots. 

Reviewer: ```suggestion
The next important part is the `leasePeriods` that corresponds to the current `auctionIndex`. We will use these `leasePeriods` to compare the winning results and see which `paraId`'s took which slots. (Note that this is a redundant way to find the auction winners when coupled with with watching for the `Leased` events)
```

> By this point we have the below relationship. 

Reviewer: Confused by the term relationship here. I think it is alluding to the fact that we can check the lease periods for a paraId and check the lease periods for an old auction to see if they overlap. Then if they overlap we know the paraId won the overlapping ones from the auction. 

However, now that I think about it, using the `Leased` events from the ending block is better is because there may be multiple auctions that cover overlapping lease periods (so testing for the overlap i mention could be misleading)

Given that, can we drop the `Relationships` section above and just recommend watching for events at the ending block?

Author: I definitely like this. It adds for a more open ended implementation while having instructions on what to watch for, and where.

> EX: (This is just an example, format the data from the endpoint however is necessary)

Reviewer: > format the data from the endpoint however is necessary

what does this mean?

Author: This => Code snippet below. 

Just changed it to `(The below code snippet is just an example, format the data from the endpoint however is necessary)`

    ],
    finishEnd: '200'
}
```

`/blocks/:blockId` ENDPOINT

Once the auction is over, its time to query the `blocks` endpoint at the block height given in the `finishEnd` field, and retrieve the events inside of `on_initialize`.

Example Response
```
{
    authorId: ....,
    extrinsics:....
    ...
    on_initialize: {
        events: [
            {
                "method": {
                    "pallet": "slots",
                    "method": "Leased"
                },
                "data": [
                    '1000', // ParaId
                    '5HpG9w8EBLe5XCrbczpwq5TSXvedjrBGCwqxK1iQ7qUsSWFc', // AccountId
                    '1', // LeasePeriod (begining of the lease period)
                    '4', // LeasePeriod (the count of the lease period)
                    '10000', // Balance (extra balance reserved)
                    '1000000000', // Balance (total balance) 
                ]
            },
            {
                "method": {
                    "pallet": "auctions",
                    "method": "AuctionClosed"
                },
                "data": [
                    ...
                ]
            }
        ]
    }
}
```

> Now that you have all the paraId's that won slots for that auction, you can compare it with the data relavant to the `auctionIndex`. Comparing the `leasePeriod`'s that are available during the active auction, to the `leasePeriod`'s that have been won and denoted in the `Leased` events (there may be multiple if there are multiple winners) will give you all the winners for that auction. 

Reviewer 2: ```suggestion
Now that you have all the `paraId`s that won slots for that auction, you can compare it with the data relevant to the `auctionIndex`. Comparing the `leasePeriod`s that are available during the active auction to the `leasePeriod`s that have been won and denoted in the `Leased` events (there may be multiple if there are multiple winners) will give you all the winners for that auction. 
```
